File: javascript/bird-watcher/bird-watcher.js
// @ts-check
//
// The line above enables type checking for this file. Various IDEs interpret
// the @ts-check directive. It will give you helpful autocompletion when
// implementing this exercise.

/**
 * Calculates the total bird count.
 *
 * @param {number[]} birdsPerDay
 * @returns {number} total bird count
 */
export function totalBirdCount(birdsPerDay) {
  return birdsPerDay.reduce((acc, value) => acc + value, 0);
}

const DAYS_PER_WEEK = 7;

/**
 * Calculates the total number of birds seen in a specific week.
 *
 * @param {number[]} birdsPerDay
 * @param {number} week
 * @returns {number} birds counted in the given week
 */
export function birdsInWeek(birdsPerDay, week) {
  return totalBirdCount(
    birdsPerDay.slice(
      week * DAYS_PER_WEEK - DAYS_PER_WEEK,
      week * DAYS_PER_WEEK
    )
  );
}

/**
 * Fixes the counting mistake by increasing the bird count
 * by one for every second day.
 *
 * @param {number[]} birdsPerDay
 * @returns {number[]} corrected bird count data
 */
export function fixBirdCountLog(birdsPerDay) {
  for (let i = 0; i < birdsPerDay.length; i++) {
    if (i % 2 == 0) birdsPerDay[i]++;
  }

  return birdsPerDay;
}

File: go/bird-watcher/bird_watcher.go
package birdwatcher

const dayPerWeek = 7

// TotalBirdCount return the total bird count by summing
// the individual day's counts.
func TotalBirdCount(birdsPerDay []int) (sum int) {
	for _, birds := range birdsPerDay {
		sum += birds
	}

	return sum
}

// BirdsInWeek returns the total bird count by summing
// only the items belonging to the given week.
func BirdsInWeek(birdsPerDay []int, week int) int {
	birdsForWeek := birdsPerDay[(week-1)*dayPerWeek : (week-1)*dayPerWeek+dayPerWeek]
	return TotalBirdCount(birdsForWeek)
}

// FixBirdCountLog returns the bird counts after correcting
// the bird counts for alternate days.
func FixBirdCountLog(birdsPerDay []int) []int {
	for i, birds := range birdsPerDay {
		if i%2 == 0 {
			birdsPerDay[i] = birds + 1
		}
	}

	return birdsPerDay
}
